### PR TITLE
Templated Filter::create

### PR DESCRIPTION
**Pull request by pshriwise**
As I was working with the C++ interface in an other project I noticed that while the `Filter::create()` method is really handy for creating these objects and automatically populating the internal data structures, the only way to interact with `Filter` sub-classes like `CellFilter` or `MeshFilter` is to re-cast the returned pointer, which is a little messy.

This adds a template overload of  `Filter::create()` so that the following:

```cpp
auto cell_filter = dynamic_cast<CellFilter*>(Filter::create("cell"));
```

can become:

```cpp
auto cell_filter = Filter::create<CellFilter>();
```

Which is a little nicer to look at and doesn't require a check that the re-cast was successful.

Along with this comes a test based on @makeclean's dynamic library source test which implement's a simple C++ driver to test this new method.

### CODE REVIEW

File: tests/regression_tests/cpp_driver/test.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            openmc.run(openmc_exec=self.executable,
+                       mpi_args=mpi_args,
+                       event_based=config['event'])
+        else:
+            openmc.run(openmc_exec=self.executable,
+                       event_based=config['event'])
+
+
+def test_cpp_driver(cpp_driver):
+    model = openmc.examples.pwr_pin_cell()
```
Comment (paulromano) [inline tests/regression_tests/cpp_driver/test.py:68]: In the interest of keeping test times down (much of which is dominated by loading cross sections), can you use a simpler model with fewer nuclides?

File: src/tallies/filter.cpp [hunk 2]
```diff
@@ -67,6 +69,20 @@ Filter::~Filter()
...
+                "Type specified is not derived from openmc::Filter");
+  // Create filter and add to filters vector
+  auto filter = std::make_unique<T>();
+  auto ptr_out = filter.get();
+  model::tally_filters.emplace_back(std::move(filter));
+  // Assign ID
+  model::tally_filters.back()->set_id(id);
+
+  return ptr_out;
+}
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply src/tallies/filter.cpp:84]: I'm cool with that -- just didn't want to break any external code. There are a couple of external applications relying on the development version of OpenMC that are using it right?

File: tests/regression_tests/cpp_driver/test.py (hunk 1 above)
Comment (pshriwise) [reply tests/regression_tests/cpp_driver/test.py:68]: Good call. Can do!

File: tests/regression_tests/cpp_driver/test.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                       mpi_args=mpi_args,
+                       event_based=config['event'])
+        else:
+            openmc.run(openmc_exec=self.executable,
+                       event_based=config['event'])
+
+
+def test_cpp_driver(cpp_driver):
+
+    model = openmc.model.Model()
```
Comment (paulromano) [inline tests/regression_tests/cpp_driver/test.py:70]: In other tests, what I've done is setup the model as a fixture:
```Python
@pytest.fixture
def model():
    model = openmc.model.Model()
    ...
    return model

def test_blah(model):
     harness = WhateverTestHarness(model)
     harness.main()

File: src/tallies/filter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulromano) [reply src/tallies/filter.cpp:84]: Yeah, I'd say just get rid of it. We are the ones working on these external applications, so not too worried about disruption here.

File: tests/regression_tests/cpp_driver/test.py (hunk 3 above)
Comment (pshriwise) [reply tests/regression_tests/cpp_driver/test.py:70]: Yeah, I'd seen that. For the `cpp_driver` it seemed to make sense as there is cleanup and teardown involved, but for the model it didn't seem like there was a need. Keeping a common pattern across tests makes them easier to navigate though, I'm sure. Updating shortly.

File: src/tallies/filter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pshriwise) [reply src/tallies/filter.cpp:84]: Ok. Away it goes!

[the remark this answers is not in the document]
Comment (pshriwise) [reply src/tallies/filter.cpp:84]: Ah, I just noticed that the `Filter::create(pugi::xml_node node)` relies on that version of the method.

**Pull request merged**